tests/test_constraints.py

```diff
@@ -99,3 +127,208 @@ def fast(self, auxdata, pars):
     slow_result = pyhf.tensorlib.tolist(slow(m, auxd, pars))
     fast_result = pyhf.tensorlib.tolist(fast(m, auxd, pars))
     assert pytest.approx(slow_result) == fast_result
+
+
+def test_ok(backend):
+    c = MockConfig(
+        par_order=['pois1', 'pois2', 'norm1', 'norm2'],
+        par_map={
+            'pois1': {
+                'paramset': constrained_by_poisson(
+                    n_parameters=1,
+                    inits=[1.0],
+                    bounds=[[0, 10]],
+                    auxdata=[12],
+                    factors=[12],
+                ),
+                'slice': slice(0, 1),
+                'auxdata': [1],
+            },
+            'pois2': {
+                'paramset': constrained_by_poisson(
+                    n_parameters=2,
+                    inits=[1.0] * 2,
+                    bounds=[[0, 10]] * 2,
+                    auxdata=[13, 14],
+                    factors=[13, 14],
+                ),
+                'slice': slice(1, 3),
+            },
+            'norm1': {
+                'paramset': constrained_by_normal(
+                    n_parameters=2,
+                    inits=[0] * 2,
+                    bounds=[[0, 10]] * 2,
+                    auxdata=[0, 0],
+                    sigmas=[1.5, 2.0],
+                ),
+                'slice': slice(3, 5),
+            },
+            'norm2': {
+                'paramset': constrained_by_normal(
+                    n_parameters=3,
+                    inits=[0] * 3,
+                    bounds=[[0, 10]] * 3,
+                    auxdata=[0, 0, 0],
+                ),
+                'slice': slice(5, 8),
+            },
+        },
+    )
+    sugg = [1.0] * 3 + [0.0] * 5  # 2 pois 5 norm
```

Comment:
If I can make a suggestion, it might slightly simplify the tests here to create the following function

```python

    def _batched_constraint(config, init, constraint_callable, batch_size=None):
        constraint = constraint_callable(config, batch_size)
        return constraint.logpdf(config.auxdata, init)
```
which then allows this first test to be

```python
    suggested_init = [1.0] * 3 + [0.0] * 5  # 2 pois 5 norm
    result = default_backend.astensor(
        pyhf.tensorlib.tolist(
            _batched_constraint(c, suggested_init, poisson_constraint_combined)
        )
    )
```

and saves some common code being repeated, such that you can also later do

```python
    sugg = [1.1] * 3 + [0.0] * 5  # 2 pois 5 norm
    result = _batched_constraint(
        c, [sugg] * 10, poisson_constraint_combined, batch_size=10
    )
    assert result.shape == (10,)
```

If you like the idea I have a commit ready to go that implements this across the test.

Reply:
Sorry, read the PR better and now see those should have been `pars` and not `init`. Changed that in the example here.

Reply:
generally I'm leaning towards having more verbose but explicit code in tests instead of logic, but I'm fine either way. If you think it's more readable and have a commit ready, go for it

Reply:
If this is your preferred style for tests I'm okay with that. Though if you don't mind I will change this commit to be more explicit in the variable naming, as I found that I had misread what was actually happening the first time through.

Reply:
> > I agree the tests become unwieldy, but i'm not sure how to resolve this, you need to test a ton of functionality. But I agree it is hard maintain, maybe we can refactor them better. At the same time I like explicitly coded out tests that don't rely on some other code, since you explicitly say what setup you are testing, otherwise the tests just become more logic to maintain
> 
> Yeah, only issue is that when a test is long and has more text, it's harder for new users to contribute (and we start getting into copy/pasting code...)

Ah, I see now that you and @kratsg already had some discussion on this. Giordon maybe you want to give some additional comments here as well before I push anything.

Reply:
I prefer more verbose tests when one can't simplify it without losing clarity.

Reply:
It seems that you're in agreement so let's keep it as is.